### PR TITLE
Added "DeterministicPolynomial" func, set RandomPolynomial to use it

### PR DESCRIPTION
**Pull request by cathalgarvey**
Hi folks,

I'm working on something that will use chunker, and I'd like to be able to deterministically generate Polynomials from secure pseudorandom input (I'll be using SHAKE256).

The changes here are self-explanatory; instead of hardcoding rand.Source I'm accepting a reader, and the `RandomPolynomial` function now simply passes rand.Source to the DeterministicPolynomial function.

Also, my linter was complaining about the dangling err in the tests, so I fatal'd it; hope you don't mind. :)

Current test suite passes, but I haven't made tests for the new DeterministicPolynomial function yet.

### CODE REVIEW

**Comment by fd0** [comment]
Hi, thanks for your contribution. Could you describe what you're trying to achieve, i.e. what's the greater picture? While reading your description I was asking myself: If you're generating a chunker polynomial in a deterministic way, why re-generate the same polynomial over and over again? Generating it once and using it afterwards seems so much easier.

In general, I'm okay with having such a function, but the description should be clearer. Let me fix the integration tests, then I'll add a proposal.

Thanks for adding the error check, that's an embarrassing bug...

File: polynomials.go [hunk 1]
```diff
@@ -160,11 +161,20 @@ const randPolMaxTries = 1e6
...
 func RandomPolynomial() (Pol, error) {
+	return DeterministicPolynomial(rand.Reader)
+}
+
+// DeterministicPolynomial uses a source of bytes to generate a polynomial.
+// The source should appear pseudorandom, at least, and by default would be
+// rand.Reader. For a secure deterministic polynomial, it could instead
+// be the output of the SHA3/SHAKE algorithm after seeding securely with
+// secret input.
+func DeterministicPolynomial(source io.Reader) (Pol, error) {
```
Comment (fd0) [inline polynomials.go]: I don't like the documentation for this function: What exactly does "appear pseudorandom" mean? It's not random, not even pseudo-random, but only "appears" to be?

I'd like to suggest naming this function `DerivePolynomial(source io.Reader)`, with a description along the lines of "Returns an irreducible Polynomial generated by reading random bytes from source", followed by the documentation for the `RandomPolynomial` function.

File: polynomials.go [hunk 2]
```diff
@@ -160,11 +161,20 @@ const randPolMaxTries = 1e6
 // "Fingerprinting by Random Polynomials", page 4. If no polynomial could be
 // found in one million tries, an error is returned.
 func RandomPolynomial() (Pol, error) {
```
Comment (fd0) [inline polynomials.go]: The documentation here should be changed to something like "Returns a random Polynomial by calling `DerivePolynomial()` with the source set to `rand.Reader`". What do you think?

**Comment by fd0** [comment]
I've fixed the integration tests, please rebase this PR against master.

**Comment by cathalgarvey** [comment]
Hey @fd0, thanks for the feedback. So, I'll:

1. Rename the function `DerivePolynomial(source io.Reader) (Pol, error)`
2. Make the documentation checks you requested.

When you say "rebase against master", can you explain? Given the simplicity here I didn't work on a feature branch, I just added it to master. Do you mean "make this a single commit's worth of changes"? Because that, I can do. :)

**Comment by cathalgarvey** [comment]
Herp, forgot to pull before making those changes and rebasing. Be gentle, that's my first time using the rebase tool and I've already got regrets. :)

**Comment by cathalgarvey** [comment]
By the way, now that I'm finished the code end, you asked what the goal is:

Right now, to use this in a repeatable fashion would require storing a static polynomial with a project that intends to re-use that polynomial consistently. That's what restic does, and it's fine for restic. But, I'd like to have consistent chunking among many separate clients, and the fewer things I need those clients to synchronise the better. In fact, I'm building architecture so that only a passphrase or keyfile needs to be synched, and all other software features derive from that..including chunking.

So, having a DerivePolynomial function means each client can generate the shared polynomial at start-up time and re-use it as needed.

Thanks for this code, I'm a daily user of Restic and I've learned a lot from it. :)

**Comment by fd0** [comment]
I understand, thanks for your contribution and the explanation!

**Comment by cathalgarvey** [comment]
Thanks Alexander!

**Pull request merged**